sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSink.scala

```diff
@@ -40,17 +41,31 @@ object FileStreamSink extends Logging {
    * be read.
    */
   def hasMetadata(path: Seq[String], hadoopConf: Configuration, sqlConf: SQLConf): Boolean = {
-    path match {
-      case Seq(singlePath) =>
-        val hdfsPath = new Path(singlePath)
-        val fs = hdfsPath.getFileSystem(hadoopConf)
-        if (fs.isDirectory(hdfsPath)) {
-          val metadataPath = getMetadataLogPath(fs, hdfsPath, sqlConf)
-          fs.exists(metadataPath)
-        } else {
-          false
-        }
-      case _ => false
+    if (sqlConf.getConf(SQLConf.FILE_SINK_FORMAT_CHECK_ENABLED)) {
```

Comment:
I think this can be simplified after checking "`isDirectory()` and `exists()` work with glob path?".

If these methods don't support glob path (and I expect they don't), we can check glob path in prior instead of dealing with exception. That would also removes the necessary of context to know "these methods can fail when the path is glob and the path is too long".

Reply:
It's a really reasonable concern. Let me share more details about the context:
- It's a regression after SPARK-26824 (you may find after the fix we move the `isDirectory` and `exists` checking out of try catch block). In our user case, the exception was thrown only with the long glob path. The same code passed in Spark 2.4.
- The exception was thrown of the checking `isDirecoty`. However, we might not be sure that it's always the same behavior since the `FileSystem` has different implementation in different systems (we met this in a non-hdfs file system).

So based on the two points above, I just chose a safer way to fix this issue. The current fix, a new config and dealing with the glob path exception, should carefully back to the behavior before SPARK-26824, and without new behavior changes.

Reply:
Let's focus on "glob path" here - from the quick look on org.apache.hadoop.fs.FileSystem javadoc, both `isDirectory()` and `exists()` seem to require the exact path, not glob path. The method which accept glob path is `globStatus()`, and there the API clearly names the parameter as `pathPattern`.

https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/fs/FileSystem.html

And intuitively, it sounds me as very odd if someone can say the glob path is a directory while matching paths can be both files and directories, and same for checking existence as well. I don't think it's feasible to "expect" the meaningful value from calling these methods with glob path. That sounds to me as "undefined behavior" even any weird file system could return true for the case.

That said, I'd rather consider the input of glob path as "wrong one" and always return false (some sort of debug log message is fine if we really like to log). If there's a code relying on such behavior, I don't think that is correct. I'd rather say the possible paths should be populated before, and this method should be called per each path.

I still need to hear voices from others, but if the consensus goes to just disallow the glob path here, we won't need to introduce the new configuration.

@tdas @zsxwing @jose-torres @viirya @gaborgsomogyi Would like to hear your voices on this. Thanks!

Reply:
cc. @steveloughran @sunchao as they could provide great inputs on Hadoop filesystem's point of view.

The summary of comment thread - is the behavior defined on providing "glob path" to `isDirectory()` and `exists()`? If defined, what is the expected behavior?

Reply:
No `isDirectory` and `exists` don't accept glob path, and it will just say "File does not exist". Internally they all call `getFileStatus` which returns information for a _single_ file, e.g., file length, permission, is it a directory or file, etc.

Reply:
@sunchao 
Thanks for the answer! That is same as my expectation.

@xuanyuanking 
According to the answer, I don't think there's a risk on dealing glob path as wrong input and just returning false. It has been working like so, plus possibility to throw additional exception. I don't think we need to add flag to keep the behavior - we can just fix it.

Reply:
Only FS API call which does glob expansion is globStatus. I don't really think that API should have been in the FS really -it's more a layer above- but it's there and has the treewalking scale problems on cloud stores to match. But, it's there.... Everything else takes the path as is.
Even there, handling of "non valid" chars in a path is trouble. There's consensus that "/" is a forbidden filename, but how to handle ":"? Long standing source of pain

Reply:
Great thanks @sunchao @steveloughran for the great inputs! Thanks @HeartSaVioR for the help!
```
Internally they all call getFileStatus which returns information for a single file
```
@sunchao That's right. The issue we are facing is the non-hdfs file system implementation of `getFileStatus` throws an exception for long globbing format. However, before SPARK-26824, Spark caught all exceptions here and just return false. After SPARK-26824, the same code failed and the user complained it's a regression.

@steveloughran yep, that's the point. The problematic one is a cloud store's implementation of a FileSystem...

Reply:
@HeartSaVioR I will change this PR for only resolving the glob path issue as our discussion. (This should be my fault to mix 2 things together.)
I'll create another PR for the behavior changing caused by SPARK-26824.

Reply:
OK. So let's sort it out:

1) filter out the glob path first and simply return false as an "early-return"
2) restore try-catch to let method tolerate the non-fatal exception

I'm OK to either doing both in here, or one by one.

Reply:
Thanks for the summary. I addressed in d7abb94.
Also, I tried to add the config in the exception branch and error message.

Reply:
@xuanyuanking 
Sorry I didn't realize you've updated this PR.

As I said earlier, if we want to ignore metadata for file sink output directory, that should be added to "source option". Let's deal with another JIRA issue / PR so that we can continue discussion afterwards.

Reply:
```
Let's deal with another JIRA issue / PR so that we can continue discussion afterwards.
```
Yep, thanks for the advice, done in 11f6bf0.
(actually that's what I mean in `I'll create another PR for the behavior changing caused by SPARK-26824`. I think I didn't express clear enough...)